### ISSUE TITLE
[MPS] pytorch nightly lost SDPA support?

### ISSUE THREAD

**Issue by bghira** [pytorch/pytorch#134678]
### 🐛 Describe the bug

On 2.4.0 we have an issue with corrupted results but in 2.5 where the corruption bug is fixed, training diffusion models no longer works due to a now-missing core component of SDPA for MPS:

```
derivative for aten::_scaled_dot_product_attention_math_for_mps is not implemented
```

@skotapati @malfet this seems like a pretty substantial regression. was it intentional? it makes MPS useless with pytorch.

### Versions

Collecting environment information...
PyTorch version: 2.5.0.dev20240828
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.0 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.30.2
Libc version: N/A

Python version: 3.10.14 (main, Mar 19 2024, 21:46:16) [Clang 15.0.0 (clang-1500.3.9.4)] (64-bit runtime)
Python platform: macOS-15.0-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] open_clip_torch==2.26.1
[pip3] torch==2.5.0.dev20240828
[pip3] torch-optimi==0.2.1
[pip3] torch-stoi==0.2.1
[pip3] torchaudio==2.5.0.dev20240828
[pip3] torchmetrics==1.4.1
[pip3] torchsde==0.2.6
[pip3] torchvision==0.20.0.dev20240828
[conda] No relevant packages

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by malfet** [comment]
Assigning to myself to understand what had happened, and why decom is not there. 

@bghira do you have a small reproducer?

**Comment by bghira** [comment]
unfortunately i havent had the ability to figure out even what line it occurs on yet because i am swamped with other projects but i will try and get more information for you soon

**Comment by qqaatw** [comment]
So I think the root cause is that the fast path added here doesn't implement its corresponding backward function: https://github.com/pytorch/pytorch/pull/131362.

I think a quick way to address it is to check whether the input tensors require grad. If required, then we don't get into the fast path and fall back to composite implicit.

**Comment by qqaatw** [comment]
I'm working on a PR to fix it

**Comment by bghira** [comment]
great! that was super fast! much impressed. thank you.

**Comment by huydhn** [comment]
Confirm that this is fixed on 2.5.0 by running `pytest -v test/test_mps.py -k test_sdpa_no_mask_no_causal_fp32_grad`
